### PR TITLE
Add an option to error on building system libs

### PR DESCRIPTION
**Pull request by kripken**
This lets us say that only the `build` stage is allowed to build system libraries, and we get a clear error if one of the later stages does. The benefit is that we can avoid duplicate work on the bots. In addition, this found a few bugs, things that embuilder doesn't yet build or builds incorrectly in ways that prevent caching.

In practice this defines a new env var, `EMCC_ERROR_ON_BUILDING_SYSTEM_LIBS`. I know @dschuff  doesn't like env vars :) but this is the kind of thing they seem appropriate for - it's convenient to set them on the circle build stages, etc., I think?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ defaults: &defaults
       environment:
         LANG: C.UTF-8
         EMTEST_DETECT_TEMPFILE_LEAKS: 1
+        EMCC_ERROR_ON_BUILDING_SYSTEM_LIBS: 1 # disallow system lib building in general - only the initial build should do it, for everyone
```
Comment (juj) [inline .circleci/config.yml]: Would a better name be `EMCC_NO_ON_DEMAND_SYSTEM_LIB_BUILDS` or something like that? `ERROR_ON_BUILDING_SYSTEM_LIBS` makes this read like `ERROR_ON_UNDEFINED_SYMBOLS` or `ERROR_ON_MISSING_LIBRARIES`, i.e. as if it was an error if system libs build fails (but otherwise might just have been a warning if system lib build fails?)

Comment (juj) [reply .circleci/config.yml]: Or alternatively `EMCC_BUILD_SYSTEM_LIBS_ON_DEMAND=0` to make the option read positive instead of negative? (undefined meaning ==1)

File: embuilder.py [hunk 2]
```diff
@@ -109,17 +102,21 @@
...
-    'dlmalloc_debug_noerrno', 'dlmalloc_threadsafe_debug_noerrno',
-    'dlmalloc_noerrno', 'dlmalloc_threadsafe_noerrno',
+    'al', 'compiler-rt', 'gl', 'gl-mt', 'gl-emu', 'libc', 'libc-mt', 'libc-extras',
+    'emmalloc', 'emmalloc_debug', 'pthreads', 'libc++', 'libc++_noexcept',
     'libc++abi', 'html5'
 ]
+for debug in ['', '_debug']:
+  for noerrno in ['', '_noerrno']:
+    for threadsafe in ['', '_threadsafe']:
+      for tracing in ['', '_tracing']:
```
Comment (dschuff) [inline embuilder.py]: For specializations that only affect one library such as dlmalloc, can we just have the user manually do `-ldlmalloc_tracing` instead of using a `-s` flag?

Comment (dschuff) [reply embuilder.py]: Oh, actually nevermind that's orthogonal to this file which just controls which flavors get built.

File: .circleci/config.yml (hunk 1 above)
Comment (dschuff) [reply .circleci/config.yml]: I like `EMCC_BUILD_SYSTEM_LIBS_ON_DEMAND`. Would also make it easier to make 0 be the default in the future if we want to pre-distribute all the libs in the SDK.

Comment (kripken) [reply .circleci/config.yml]: Sounds good, renaming.

**Comment by kripken** [comment]
Ok, changing to a settings file option.

**Comment by kripken** [comment]
Oh, wait, this won't quite work as a settings option. We'd need to set it differently in each build on CI, that is, write to the settings file. But modifying the settings file will clear the cache... but we'd be preventing rebuilding those things.

Perhaps this could be a settings option, and it would also disable ever clearing the cache? That seems like the right behavior, in a way, but also adds complexity.

**Comment by dschuff** [comment]
I guess we could make it a config file option, but then have an env var to suppress clearing the cache? Then the env var would only ever be needed on CI where we want to test one and then the other, right? Or is that just too crazy?

**Comment by kripken** [comment]
I like the part about using a config file option, but adding an env var to suppress clearing the cache is I think a weirder option than the main option proposed here. Feels like we should figure out something better for when to clear the cache, if we want to go that way - perhaps the heuristic of "settings file changed, so best to clear the cache" is itself something we should reconsider.

**Comment by kripken** [comment]
If we are not happy with adding this env var, I can land all the other parts of this PR (the fixes found by it).

File: .circleci/config.yml [hunk 3]
```diff
@@ -4,6 +4,7 @@ defaults: &defaults
       environment:
         LANG: C.UTF-8
         EMTEST_DETECT_TEMPFILE_LEAKS: 1
+        EMCC_BUILD_SYSTEM_LIBS_ON_DEMAND: 0 # disallow system lib building in general - only the initial build should do it, for everyone
```
Comment (dschuff) [inline .circleci/config.yml]: These can go now, right?

File: tests/runner.py [hunk 4]
```diff
@@ -196,6 +196,33 @@ def modified(self):
   return decorated
 
 
+@contextlib.contextmanager
+def shared_modify(updates):
+  """A context manager that updates shared."""
```
Comment (dschuff) [inline tests/runner.py]: this text is a bit confusing at first glance but I suppose it's just because of the existing naming convention.

File: .circleci/config.yml (hunk 3 above)
Comment (kripken) [reply .circleci/config.yml]: yes, just removed now.

File: tests/runner.py (hunk 4 above)
Comment (kripken) [reply tests/runner.py]: Sorry, still working on this - I was wrong, this can't work, so I removed it.

When tests are passing again I'll comment with a summary of the new changes.

**Comment by kripken** [comment]
Hopefully tests will pass now.

Summary of what this PR does, after rethinking things and offline discussion:

 * Add a FROZEN_CACHE option to the settings file. As the name implies, this means the cache is not to be changed. As a result, the cache is not cleared (but we do warn if we think it should be), and system libraries are not built into it, instead we error.
 * Add missing libraries to embuilder.py, that error checking on the bots found.
 * Somewhat regrettably, some tests had to be moved from `other` to `sanity` - anything which fundamentally builds stuff to the cache has to be there, since `sanity` is the only place we allow building stuff now. The regrettable part is that sanity is not fast already. Perhaps this makes it more urgent to look into streamlining it.
 * A few misc fixes related to this, like cocos2d had an inconsistent naming of its files, which made it confusing to debug errors with it not being built.

File: tests/test_other.py [hunk 5]
```diff
@@ -7495,6 +7445,7 @@ def test_function_eliminator_replace_variable_value(self):
     self.function_eliminator_test_helper('test-function-eliminator-replace-variable-value.js',
                                          'test-function-eliminator-replace-variable-value-output.js')
 
+  @no_wasm_backend('tests asm.js optimizer')
```
Comment (sbc100) [inline tests/test_other.py]: How was this working before?

File: tests/test_other.py [hunk 6]
```diff
@@ -8131,25 +8082,6 @@ def test_wasm_targets_side_module(self):
...
-      assert Building.is_wasm('a.o') and not Building.is_bitcode('a.o')
-      print('bitcode in object')
-      run_process([PYTHON, EMCC, path_from_root('tests', 'hello_libcxx.cpp')] + args + ['-c', '-o', 'a.o', '-s', 'WASM_OBJECT_FILES=0'])
-      assert not Building.is_wasm('a.o') and Building.is_bitcode('a.o')
-      print('build bitcode object')
-      run_process([PYTHON, EMCC, 'a.o'] + args + ['-s', 'WASM_OBJECT_FILES=0'])
-      self.assertContained('hello, world!', run_js('a.out.js'))
-      print('build with bitcode')
-      run_process([PYTHON, EMCC, path_from_root('tests', 'hello_libcxx.cpp')] + args + ['-s', 'WASM_OBJECT_FILES=0'])
-      self.assertContained('hello, world!', run_js('a.out.js'))
```
Comment (sbc100) [inline tests/test_other.py]: Why did this test end up moving?     It doesn't look like it should be writing to the cache.  Do we cache bitcode libraries and object file library separately?   If so we should we build the bitcode ones as part of the embuiilder phase too, no?

File: tests/test_other.py (hunk 5 above)
Comment (kripken) [reply tests/test_other.py]: It did work, as the test just runs the optimizer directly. But it's redundant (nothing is done differently than the fastcomp path), so I disabled it here.

File: tests/test_other.py (hunk 6 above)
Comment (kripken) [reply tests/test_other.py]: Yes, we have a separate dir for bitcode and object files in the cache. And embuilder doesn't currently have a way to build both, so it just builds the default object files now. So this is the one test which builds the LTO libraries.

Building the LTO ones by default seems costly in terms of build time, but yeah, maybe that's the best thing to avoid confusion.

Comment (sbc100) [reply tests/test_other.py]: You mean the build time on the circleCI bots?   Slowing down the time it takes to run embuilder?  It would slow down only the `build-upstream` build step right?   Maybe double the time this step takes.  I think this is reasonable, and it won't effect normal users or developers local build times.

Comment (kripken) [reply tests/test_other.py]: Just the time it takes to run embuilder, yeah. But normal users may be running embuilder too, and this is a significant slowdown. I think it may be the right way to go, but I'd suggest discussing it in a dedicated issue, I opened https://github.com/emscripten-core/emscripten/issues/8128

Comment (kripken) [reply tests/test_other.py]: Moved this test back after landing the `--lto` flag for embuilder.

File: tests/test_sanity.py [hunk 7]
```diff
@@ -501,6 +503,71 @@ def ensure_cache():
...
+    # FROZEN_CACHE prevents cache clears, and prevents building
+    ensure_cache()
+    self.assertTrue(os.path.exists(EMCC_CACHE))
+    self.assertTrue(os.path.exists(Cache.root_dirname))
+    add_to_config('FROZEN_CACHE = True')
+    self.do([PYTHON, EMCC])
+    self.assertTrue(os.path.exists(EMCC_CACHE))
+    self.assertTrue(os.path.exists(Cache.root_dirname))
+    output = self.do([PYTHON, EMBUILDER, 'build', 'emmalloc'])
+    self.assertIn('FROZEN_CACHE disallows building system libs', output)
```
Comment (sbc100) [inline tests/test_sanity.py]: Can you make this a separate test?  This test is super long already.

File: tools/system_libs.py [hunk 8]
```diff
@@ -441,20 +441,20 @@ def require_dlmalloc(what):
...
+        extra += '_noerrno'
     if shared.Settings.USE_PTHREADS:
       extra += '_threadsafe'
-      require_dlmalloc('pthreads')
     if shared.Settings.EMSCRIPTEN_TRACING:
       extra += '_tracing'
       require_dlmalloc('tracing')
-    if shared.Settings.DEBUG_LEVEL >= 3:
-      extra += '_debug'
+      require_dlmalloc('pthreads')
```
Comment (sbc100) [inline tools/system_libs.py]: Shouldn't this line be after 451?

If so I guess this shows we are not testing this.

Comment (kripken) [reply tools/system_libs.py]: Oops, yeah, thanks.

And yeah, we don't test emmalloc errors on these corner cases I guess. emmalloc is not a default option so it's gotten less attention I guess.
